openmls/src/framing/public_message_in.rs

```diff
@@ -0,0 +1,325 @@
+        }
+    }
+}
+
+// === Helper structs ===
+
+/// 9.2 Transcript Hashes
+///
+/// ```c
+/// // draft-ietf-mls-protocol-16
+///
+/// struct {
+///    WireFormat wire_format;
+///    FramedContent content; /* with content_type == commit */
+///    opaque signature<V>;
+///} ConfirmedTranscriptHashInput;
+/// ```
+#[derive(TlsSerialize, TlsSize)]
+pub(crate) struct ConfirmedTranscriptHashInput<'a> {
+    pub(super) wire_format: WireFormat,
+    pub(super) mls_content: &'a FramedContentIn,
+    pub(super) signature: &'a Signature,
+}
+
+#[derive(TlsSerialize, TlsSize)]
+pub(crate) struct InterimTranscriptHashInput<'a> {
+    pub(crate) confirmation_tag: &'a ConfirmationTag,
+}
+
+impl<'a> TryFrom<&'a PublicMessageIn> for InterimTranscriptHashInput<'a> {
+    type Error = &'static str;
+
+    fn try_from(public_message: &'a PublicMessageIn) -> Result<Self, Self::Error> {
+        match public_message.auth.confirmation_tag.as_ref() {
+            Some(confirmation_tag) => Ok(InterimTranscriptHashInput { confirmation_tag }),
+            None => Err("PublicMessage needs to contain a confirmation tag."),
+        }
+    }
+}
+
+impl<'a> From<&'a ConfirmationTag> for InterimTranscriptHashInput<'a> {
+    fn from(confirmation_tag: &'a ConfirmationTag) -> Self {
+        InterimTranscriptHashInput { confirmation_tag }
+    }
+}
+
+impl TlsDeserializeTrait for PublicMessageIn {
+    fn tls_deserialize<R: Read>(bytes: &mut R) -> Result<Self, tls_codec::Error> {
+        let content = FramedContentIn::tls_deserialize(bytes)?;
+        let auth = FramedContentAuthDataIn::deserialize(bytes, content.body.content_type())?;
+        let membership_tag = if content.sender.is_member() {
+            Some(MembershipTag::tls_deserialize(bytes)?)
+        } else {
+            None
+        };
+
+        Ok(PublicMessageIn::new(content, auth, membership_tag))
+    }
+}
+
+impl Size for PublicMessageIn {
+    #[inline]
+    fn tls_serialized_len(&self) -> usize {
+        self.content.tls_serialized_len()
+            + self.auth.tls_serialized_len()
+            + if let Some(membership_tag) = &self.membership_tag {
+                membership_tag.tls_serialized_len()
+            } else {
+                0
+            }
+    }
+}
+
+impl TlsSerializeTrait for PublicMessageIn {
+    fn tls_serialize<W: Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
+        // Serialize the content, not the TBS.
+        let mut written = self.content.tls_serialize(writer)?;
+        written += self.auth.tls_serialize(writer)?;
+        written += if let Some(membership_tag) = &self.membership_tag {
+            membership_tag.tls_serialize(writer)?
+        } else {
+            0
+        };
+        Ok(written)
+    }
+}
+
+// The following two `From` implementations break abstraction layers and MUST
+// NOT be made available outside of tests or "test-utils".
+
+// TODO: Re-enable this #[cfg(any(feature = "test-utils", test))]
```

Comment:
Same as for the other instance of this comment.

Reply:
Not sure what the other instance is...